nemo_aligner/algorithms/spin.py

```diff
@@ -238,38 +274,46 @@ def get_generations(self, list_of_batches):
         )
         prompt_tokens = prompt_tokens.cuda(non_blocking=True)
         prompt_lengths = prompt_lengths.cuda(non_blocking=True)
+        inputs = (prompt_tokens, prompt_lengths)
+
+        if self.use_trtllm_generation:
+            actor_output = self.trtllm_generate.generate(inputs)
+            response_tokens = actor_output["response_tokens"].cpu()
+            response_lengths = actor_output["response_lengths"].cpu()
```

Comment:
The bane of my existence. Unfortunately, TRT-LLM fixes the generation/prompt lengths during compile time when it builds the engine, and you cannot change it without rebuilding the engine all over again. So when you get into the TRT block in SPIN, your prompt must be <= max_prompt_length (so 2048 in a 4096 model), or you will get a crash. We can in theory re-compile the model to adjust for prompts which are too long, but this will add overhead so I didn't pursue that path. Nemo's model.generate() does not have this limitation, so you can dynamic change the prompt limit since it's not actually a limit, you just have fewer positions for generated tokens.

To work around this limitation, I have added a new run-time truncation feature which I am using in SPIN and Self-Rewarding in my latest codebase: https://github.com/NVIDIA/NeMo-Aligner/blob/de8dda801e4ca2b72c4075d2908e8ffeade2296d/nemo_aligner/data/nlp/datasets.py#L556

Additionally, all of the updated TRT code for SPIN/Self-Rewarding in my branch above have various workarounds to intelligentally truncate to ensure we never breach the TRT prompt length limit. For example, for Self-Rewarding we use custom templates for sending to the judge models, so you have to check those manually that you don't exceed your prompt length limit, for example here: https://github.com/NVIDIA/NeMo-Aligner/blob/4f81524ec2b1a774900f54930234972c98b19d0e/nemo_aligner/algorithms/self_rewarding.py#L903

Reply:
I see, thanks!

In that case can we explicitly crash in the Python code instead of waiting for TRT-LLM to crash? (and move `adj_generation_length` to the `else:` (i.e., non-TRT-LLM) block. I think it'll make it easier for the reader to understand what's going on (and be confident that it's not an oversight in the implementation).

Reply:
The way it's written now works correctly, you will see that `adj_generation_length` is only used inside of the else block. It's calculated outside of the whole if block, but it's only ever used in its appropriate section. Any prompt length errors do happen on the Python side, they get thrown inside of trt_llm.py inside Aligner. I've tested both extensively (TRT on/off).

Reply:
I'm not saying it doesn't work correctly, I'm just suggesting the above changes for code readability (right now when I see this code I think "wait a minute there's probably a bug here!" -- which was the reason for my original comment).

Reply:
Let's remove SPIN from this PR so it won't be an issue.

Reply:
done, c4881fa997a56bdd6032a2e1726cd4f7e3e50cae let's separate this out. I realize I should have done that in the first place

Reply:
Ok, resolving then. When we bring it back it may be a good time to look into #272 so as to factorize the code and benefit from all the shenanigans implemented for PPO.